flink-baseline/flink-jdbc-sink/src/main/java/com/ness/flink/sink/jdbc/connector/SimpleJdbcConnectionProvider.java

```diff
@@ -17,52 +17,121 @@
+        if (loadedDriver == null) {
+            loadedDriver = loadDriver(jdbcConnectionOptions.getDriverName());
+        }
+        return loadedDriver;
     }
 
     @Override
-    public Connection getConnection() throws SQLException, ClassNotFoundException {
-        if (connection == null) {
-            synchronized (this) {
-                if (connection == null) {
-                    Class.forName(jdbcOptions.getDriverName());
-                    if (jdbcOptions.isUseDbURL()) {
-                        log.info("Trying to open connection based on dbURL: {}", jdbcOptions);
-                        connection = DriverManager.getConnection(jdbcOptions.getDbURL());
-                    } else {
-                        log.info("Trying to open connection based on username: {}", jdbcOptions);
-                        connection = DriverManager.getConnection(jdbcOptions.getDbURL(), jdbcOptions.getUsername(), jdbcOptions.getPassword().orElse(null));
-                    }
-                    Optional<Boolean> autoCommit = jdbcOptions.getAutoCommit();
-                    if (autoCommit.isPresent()) {
-                        connection.setAutoCommit(autoCommit.get());
-                    }
-                }
+    public Connection establishConnection() throws SQLException, ClassNotFoundException {
+        Connection connection;
+        if (jdbcConnectionOptions.getDriverName() == null) {
+            log.debug("Opening connection. Driver wasn't provided");
+            connection = DriverManager.getConnection(
+                jdbcConnectionOptions.getDbURL(),
+                jdbcConnectionOptions.getUsername(),
+                jdbcConnectionOptions.getPassword().orElse(null));
+            setAutoCommit(connection);
+        } else {
+            log.debug("Opening connection. Driver was provided");
+            Driver driver = getLoadedDriver();
+            Properties info = new Properties();
+            if (jdbcConnectionOptions.getUsername() != null) {
+                info.setProperty("user", jdbcConnectionOptions.getUsername());
             }
+            jdbcConnectionOptions.getPassword().ifPresent(password -> info.setProperty("password", password));
+            connection = driver.connect(jdbcConnectionOptions.getDbURL(), info);
+            if (connection == null) {
+                // Throw same exception as DriverManager.getConnection when no driver found to match
+                // caller expectation.
+                throw new SQLException("No suitable driver found for " + jdbcConnectionOptions.getDbURL(), "08001");
+            }
+            setAutoCommit(connection);
         }
-        log.debug("Connection: {}", connection);
         return connection;
     }
+
+    private void setAutoCommit(Connection connection) throws SQLException {
+        Optional<Boolean> autoCommit = jdbcConnectionOptions.getAutoCommit();
+        if (autoCommit.isPresent()) {
```

Comment:
for optional you better use `getAutoCommit().ifPresent(...)`

Reply:
It will be ugly because I have to handle SQLException
```
        autoCommit.ifPresent(b -> {
            try {
                connection.setAutoCommit(b);
            } catch (SQLException e) {
                throw new RuntimeException(e);
            }
        });
```
In my version I don't have to do it

Reply:
And see how it will be
```
    private void setAutoCommit(Connection connection) throws SQLException {
        Optional<Boolean> autoCommit = jdbcConnectionOptions.getAutoCommit();
        autoCommit.ifPresent(b -> {
            try {
                connection.setAutoCommit(b);
            } catch (SQLException e) {
                throw new RuntimeException(e);
            }
        });
    }```
vs current one
```
    private void setAutoCommit(Connection connection) throws SQLException {
        Optional<Boolean> autoCommit = jdbcConnectionOptions.getAutoCommit();
        if (autoCommit.isPresent()) {
            connection.setAutoCommit(autoCommit.get());
        }
    }
```

what would you prefer? ;-)